### PR TITLE
Expand test cloud provider with autoprovisioning options

### PR DESCRIPTION
**Pull request by mwielgus**
To allow ScaleUp/ScaleDown unit test for Node Autoprovisioning.

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/test/test_cloud_provider.go [hunk 1]
```diff
@@ -250,10 +308,24 @@ func (tng *TestNodeGroup) Nodes() ([]string, error) {
...
 func (tng *TestNodeGroup) Autoprovisioned() bool {
-	return false
+	return tng.autoprovisioned
 }
 
 // TemplateNodeInfo returns a node template for this node group.
 func (tng *TestNodeGroup) TemplateNodeInfo() (*schedulercache.NodeInfo, error) {
-	return nil, cloudprovider.ErrNotImplemented
+	if tng.cloudProvider.machineTemplates == nil {
+		return nil, cloudprovider.ErrNotImplemented
```
Comment (aleksandra-malinowska) [inline cluster-autoscaler/cloudprovider/test/test_cloud_provider.go]: Does ErrNotImplemented indicate there's a plan to handle it differently in the future?

File: cluster-autoscaler/cloudprovider/test/test_cloud_provider.go [hunk 2]
```diff
@@ -170,24 +227,25 @@ func (tng *TestNodeGroup) IncreaseSize(delta int) error {
...
 // theoretical node group from the real one.
 func (tng *TestNodeGroup) Exist() bool {
-	return true
+	return tng.exist
 }
 
 // Create creates the node group on the cloud provider side.
 func (tng *TestNodeGroup) Create() error {
-	return cloudprovider.ErrAlreadyExist
+	tng.cloudProvider.AddAutoprovisionedNodeGroup(tng.id, tng.minSize, tng.maxSize, 0, tng.machineType)
```
Comment (aleksandra-malinowska) [inline cluster-autoscaler/cloudprovider/test/test_cloud_provider.go]: Can we verify here (or in AddAutoporovisionedNodeGroup) that the node group doesn't already exist first? If we just overwrite it in tests we may miss some bugs related to avoiding/handling this case.

File: cluster-autoscaler/cloudprovider/test/test_cloud_provider.go [hunk 3]
```diff
@@ -95,13 +121,22 @@ func (tcp *TestCloudProvider) Pricing() (cloudprovider.PricingModel, errors.Auto
...
 // created on the cloud provider side. The node group is not returned by NodeGroups() until it is created.
 func (tcp *TestCloudProvider) NewNodeGroup(machineType string, labels map[string]string, extraResources map[string]resource.Quantity) (cloudprovider.NodeGroup, error) {
-	return nil, cloudprovider.ErrNotImplemented
+	return &TestNodeGroup{
+		cloudProvider:   tcp,
+		id:              "autoprovisioned-" + machineType,
+		minSize:         0,
+		maxSize:         1000,
+		targetSize:      0,
+		exist:           true,
```
Comment (aleksandra-malinowska) [inline cluster-autoscaler/cloudprovider/test/test_cloud_provider.go]: Shouldn't "exist" field be set to false here, as this is a theoretical node group? Based on Exist() semantics below: "Allows to tell the theoretical node group from the real one."

Comment (mwielgus) [reply cluster-autoscaler/cloudprovider/test/test_cloud_provider.go]: This is a real (not autoprovisioned group) so it should exist.

File: cluster-autoscaler/cloudprovider/test/test_cloud_provider.go (hunk 2 above)
Comment (mwielgus) [reply cluster-autoscaler/cloudprovider/test/test_cloud_provider.go]: Done.

File: cluster-autoscaler/cloudprovider/test/test_cloud_provider.go (hunk 1 above)
Comment (mwielgus) [reply cluster-autoscaler/cloudprovider/test/test_cloud_provider.go]: no, it mimics node group that don't support templatenodeinfo and ErrNotImplemented is the most probable error returned from them.

File: cluster-autoscaler/cloudprovider/test/test_cloud_provider.go (hunk 3 above)
Comment (mwielgus) [reply cluster-autoscaler/cloudprovider/test/test_cloud_provider.go]: Fixed.

**Pull request merged**